net: l2: ieee802154: improved/fixed endianness and short address handling

fgrandel opened the pull request:
This PR implements part one of the change announced in https://github.com/zephyrproject-rtos/zephyr/issues/50336#issuecomment-1250250154:

> Improve short address support in the IEEE 802.15.4 stack as a precondition for TSCH link tables (currently short addresses are not fully supported)

It turned out that a neighbor table was not (yet) required for this preparatory change. The PR also fixes endianness in the IEEE 802.15.4 L2 stack as the work on this change revealed several inconsistencies and even bugs in byte order conversion.

Commit:
net: ip: fix minor formatting glitch

This changes just fixes an erroneous indentation from a previous patch.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: add short address to ll address

This is a preparatory change that fixes one aspect of short address
handling before fixing endianness so that the endianness fix can be
applied consistently in this method.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: improve short address support

IEEE 802.15.4 short address support is incomplete in several places.
This change improves short address support without claiming to fix
it everywhere. Future iterations will have to continue where this change
leaves off.

The purpose of this change was to:
 * use the short address returned by association responses,
 * automatically bind IEEE 802.15.4 datagram sockets to the short
   address if available,
 * use the short address in outgoing packages where applicable,
 * improve validation of association/disassociation frames,
 * model association more closely to the spec by tying it to the
   existence of a short address in the MAC PIB thereby removing
   redundancy in the PIB (which makes race conditions less probable),
 * keep both, the short and extended addresses, of the coordinator.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: improved context thread safety

Several attributes in the ieee802154_context struct may potentially be
accessed from different threads and/or ISR context. Only some of these
attributes were properly guarded against race conditions.

This may not have been to problematic in the past but as other changes
in this PR introduce additional attributes and mutate several attributes
in a single atomic transaction, leaving such changes unprotected seems
dangerous.

This change therefore introduces systematic locking of the
ieee802154_context structure.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

rlubos commented on include/zephyr/net/ieee802154_mgmt.h:
```diff
@@ -42,14 +42,14 @@ enum net_request_ieee802154_cmd {
+	NET_REQUEST_IEEE802154_CMD_GET_CHANNEL, /* in CPU byte order */
+	NET_REQUEST_IEEE802154_CMD_SET_PAN_ID, /* in CPU byte order */
+	NET_REQUEST_IEEE802154_CMD_GET_PAN_ID, /* in CPU byte order */
+	NET_REQUEST_IEEE802154_CMD_SET_EXT_ADDR, /* in big endian byte order */
```
The commit message in `net: l2: ieee802154: document and fix endianness` mentions "external address" twice - shouldn't it rather be "extended address"?

fgrandel replied:
Oups. Of course. This was just a test whether you actually read my commit messages. ;-)

rlubos commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -271,24 +271,29 @@ struct ieee802154_config {
 
 		/** ``IEEE802154_CONFIG_CSL_RX_TIME`` */
-		uint32_t csl_rx_time;
+		uint32_t
+			csl_rx_time; /* in microseconds, based on ieee802154_radio_api.get_time() */
```
Why such an unexpected line break? IMO better to break the comment if it does not fit.

fgrandel replied:
Fixed.

rlubos commented on subsys/net/ip/net_context.c:
```diff
@@ -758,17 +758,17 @@ int net_context_bind(struct net_context *context, const struct sockaddr *addr,
-				ll_addr->sll_halen;
-		} else {
+
+#ifdef CONFIG_NET_L2_IEEE802154
```
It's already within `IS_ENABLED(CONFIG_NET_L2_IEEE802154)` block, so this #ifdef seems redundant?

fgrandel replied:
Fortunately I could solve this differently now so no need to mention a preprocessor constant at all.

The problem was, that the code wouldn't compile without `#ifdef` as `IS_ENABLED()` will let the compiler see the code inside the block which only works if all code referenced inside this block is also visible to the compiler. This wasn't the case for some struct field in this block which was defined conditionally. Using `IS_ENABLED()` for struct fields wastes memory, though, that the compiler cannot optimize away AFAICS. So while I quite like the `IS_ENABLED()` pattern now that I fully understand its purpose it doesn't seem to work in such cases.

rlubos reviewed (commented):
Had a brief look, not much to complain, just a few nits. Looks pretty good!

Commit:
net: l2: ieee802154: fix frame type field check

The LLDN frame has been obsoleted in IEEE 802.15.4-2015f. This change
removes it from the code, introduces frame types from current spec
levels and updates the frame validation rules in accordance with the
spec.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

fgrandel commented:
@rlubos : Thanks for the review! I also fixed the remaining bugs today, that made the tests fail and I added extra test coverage for short address mode. So should be ready for final review.

rlubos reviewed (dismissed):
LGTM

jukkar reviewed (dismissed):
LGTM

tbursztyka commented on include/zephyr/net/ieee802154.h:
```diff
@@ -42,6 +42,8 @@ extern "C" {
 
 /* See IEEE 802.15.4-2006, section 7.2.1.4 */
 #define IEEE802154_BROADCAST_ADDRESS 0xFFFF
+#define IEEE802154_NO_SHORT_ADDRESS_ASSIGNED 0xFFFE
```
Split the commit. It's too many things in one, hard to follow.
- one about the short address it 802.15.4
- one for context
- one about the association/disassociation stuff

fgrandel replied:
Fixed.

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -763,15 +763,25 @@ int net_context_bind(struct net_context *context, const struct sockaddr *addr,
 				return -EINVAL;
 			}
 
+			struct ieee802154_context *ieee802154_context =
```
No, no l2 stuff into net_context.c
You'll have to find another way.

fgrandel replied:
This led me to better investigate the locking strategy of the ll address in net_if. I'm now relying on net_if's own lock which is much better. No need to do any L2-specific locking to update the ll-address any more. Also reminded me to cover the corresponding ieee802154_mgmt API in the test which again revealed a bug. Thanks for this useful remark.

tbursztyka commented on subsys/net/l2/ieee802154/ieee802154.c:
```diff
@@ -380,9 +387,12 @@ static int ieee802154_enable(struct net_if *iface, bool state)
 
 	NET_DBG("iface %p %s", iface, state ? "up" : "down");
 
+	k_sem_take(&ctx->ctx_lock, K_FOREVER);
```
Please add empty lines: 
- before the if (unless its statement is about what has been computed before), linux code style
- after the '}' (unless it's another '}') etc..

Apply that everywhere

fgrandel replied:
Fixed here and in all other places throughout the whole change set.

tbursztyka commented on subsys/net/l2/ieee802154/ieee802154_frame.c:
```diff
@@ -916,19 +936,24 @@ bool ieee802154_create_ack_frame(struct net_if *iface, struct net_pkt *pkt, uint
 bool ieee802154_decipher_data_frame(struct net_if *iface, struct net_pkt *pkt,
 				    struct ieee802154_mpdu *mpdu)
 {
+	bool ret = false;
```
try to keep a reversed christmas tree variable ordering when applicable.

fgrandel replied:
Fixed, but where can I find this rule (and others you claimed) documented? I looked in the [coding guidelines](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html) and my code is passing all automatic style checks. If such rules are expected by the project why are they not enforced by the style checker and/or documented? Or did I simply overlook them?

tbursztyka commented on subsys/net/l2/ieee802154/ieee802154_mgmt.c:
```diff
@@ -150,28 +157,30 @@ static int ieee802154_scan(uint32_t mgmt_request, struct net_if *iface,
 			if (ret) {
 				NET_DBG("Could not send Beacon Request (%d)",
 					ret);
-				net_pkt_unref(pkt);
```
No, the packet got referenced above: line 153.

fgrandel replied:
This code block now calls the "out" label where the packet is being unrefed. Therefore this should be ok.

tbursztyka replied:
No, it got referenced once more just 7 lines before, we need to get rid of this reference.
When the pkt gets created, its reference count has been initialized to 1. Here in this code block we put it to 2. 
If we specifically do not get back to a ref count of 1, at the out label it will get back to 1 and... the pkt will not be freed. (it needs to get to 0 for this to happen). So this pkt will be accounted as memory loss.

fgrandel replied:
As the comment was hidden I overlooked your response. Will fix this now.

tbursztyka commented on subsys/net/l2/ieee802154/ieee802154_mgmt.c:
```diff
@@ -258,26 +272,34 @@ enum net_verdict ieee802154_handle_mac_command(struct net_if *iface,
 
+		k_sem_take(&ctx->ctx_lock, K_FOREVER);
+
+		enum net_verdict ret = NET_DROP;
```
kind of ugly to see a variable declaration it a middle of a code block

fgrandel replied:
Fixed.

tbursztyka commented on include/zephyr/net/ieee802154.h:
```diff
@@ -85,11 +82,15 @@ struct ieee802154_context {
+	uint8_t ack_received : 1;  /* guarded by ack_lock */
+	uint8_t ack_requested : 1; /* guarded by ack_lock */
+	uint8_t _unused : 6;
+	struct k_sem ack_lock;
```
looks like reordering the attributes this way, you mess up the structure alignment. (thus probably lose some space)

fgrandel replied:
Good point. Not the re-ordering was problematic but the introduction of the additional semaphore in the wrong place. Fixed this. There was also additional pre-existing padding in the struct (16 bytes overall) that could be removed without impacting readability. This is done in a separate commit now.

tbursztyka reviewed (changes requested):
Overall a good PR, but some commits are too big and include too many changes at once: please split them. Side note: try to keep consistent with the code style.

fgrandel commented:
> Overall a good PR, but some commits are too big and include too many changes at once: please split them. Side note: try to keep consistent with the code style.

Thank you @tbursztyka for this thorough review. Will look into it this weekend to address your concerns.

Commit:
net: ip: test and clear net_if flag

This change introduces an additional function into the API that allows
to test and clear a net_if flag atomically. A similar function already
exists to test and set a flag. So this change actually improves symmetry
of the API.

The change is required in later commits of this PR.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
drivers: ieee802154: document endianness

The IEEE 802.15.4 radio driver encodes attributes in:

* little endian for everything that is close to the protocol as
  IEEE 802.15.4 frames are little endian encoded.

* mixed big and little endian in its configuration where extended
  addresses are being represented. These inconsistencies are unfortunate
  but cannot be easily fixed in a backwards compatible way so will be
  left untouched in this change.

Endianness was almost nowhere documented which explains these
inconsistencies and led to several bugs where assignments of different
byte order are not converted (or sometimes converted, sometimes not).

This change documents intended endianness within the realm of the
IEEE 802.15.4 radio driver code. Conversion bugs are fixed in a separate
commit.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: ip: document endianness

To maintain POSIX compatibility, link-layer addresses are encoded in big
endian in the IP stack and socket API.

The intended endianness was however not documented everywhere which led
to bugs and inconsistencies. The IEEE 802.15.4 L2 stack, for example,
sometimes stores addresses in little endian in structures that intend
them to be stored in POSIX-compliant big endian byte order.

This change documents intended endianness within the realm of the
IP and sockets stack. Conversion bugs are fixed in a separate commit.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: document endianness

The IEEE 802.15.4 L2 code stores representation of attributes like
PAN id, short address and extended address in different encodings:

* big endian for extended address and CPU byte order for everything
  else whenever such attributes enter user space (except for IP/socket
  link layer addresses which are always big endian - even in case of
  short addresses - to maintain POSIX compatibility).

* little endian for everything that is close to the radio driver as
  IEEE 802.15.4 frames are little endian encoded.

Endianness was almost nowhere documented which led to several bugs and
inconsistencies where assignments of different byte order were not
converted (or sometimes converted, sometimes not).

This change documents endianness wherever possible within the realm of
the IEEE 802.15.4 L2 code. Conversion bugs and inconsistencies that were
revealed by the improved documentation will be fixed in a separate
commit.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
samples: net: wpanusb: document endianness

The WPAN-USB sample did not document endianness of some user space
variables. As the IEEE 802.15.4 stack uses attributes in several
different encodings, the endianness should be documented.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: fix short/ext address endianness

This changes fixes several bugs and inconsistencies in the IEEE 802.15.4
L2 implementation. These bugs were revealed while documenting intended
endianness of driver, IP, socket and L2 attributes (see previous
changes).

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

fgrandel commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -940,9 +940,10 @@ static int nrf5_configure(const struct device *dev,
 		sys_put_le16(config->ack_ie.short_addr, short_addr_le);
 		/**
 		 * The extended address field passed to this function starts
-		 * with the leftmost octet and ends with the rightmost octet.
```
Fixed.

fgrandel commented on subsys/net/ip/connection.c:
```diff
@@ -760,8 +760,7 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 		}
-			/* loop end */
-	}
+	} /* loop end */
```
Really? For a one-liner? Seems like a lot of maintenance overhead for the maintainers (and myself). As an external contributor I'd stop fixing such things if they always had to be done in separate PRs. (While a separate commit is totally fine of course.) If you insist I'll do so but it would be great if you could extend a little on your rationale behind it so it becomes easier to accept this.

fgrandel commented on include/zephyr/net/net_ip.h:
```diff
@@ -201,23 +201,23 @@ struct sockaddr_in_ptr {
-	uint8_t     sll_pkttype;  /* Packet type             */
-	uint8_t     sll_halen;    /* Length of address       */
-	uint8_t     sll_addr[8];  /* Physical-layer address  */
+	sa_family_t sll_family;   /* Always AF_PACKET                   */
```
This is closely related to other stuff in this PR. So while I definitely agree that this should go into a separate commit so that maintainers can pick commits relevant to their code ownership I do not yet understand the point of creating separate PRs (but I'm of course open to change my mind with a good argument). As in the other comment above: If you insist I'll of course follow your advice but would be great if you could give the "why?" together with the "what?".

tbursztyka replied:
A split commit will do

fgrandel commented on include/zephyr/net/net_linkaddr.h:
```diff
@@ -66,7 +66,7 @@ enum net_link_type {
 struct net_linkaddr {
 	/** The array of byte representing the address */
-	uint8_t *addr;
+	uint8_t *addr; /* in big endian */
```
Duplicate, see discussion above. Went into a different commit.

fgrandel commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -313,7 +319,7 @@ struct ieee802154_radio_api {
 
 	/** Set current channel */
-	int (*set_channel)(const struct device *dev, uint16_t channel);
+	int (*set_channel)(const struct device *dev, uint16_t channel /* in CPU byte order */);
```
Fixed.

fgrandel commented on subsys/net/l2/ieee802154/ieee802154_frame.h:
```diff
@@ -114,8 +114,8 @@ struct ieee802154_fcf_seq {
 		uint8_t simple_addr;
-		uint16_t short_addr;
-		uint8_t ext_addr[0];
+		uint16_t short_addr; /* in little endian */
```
Fixed.

Commit:
net: l2: ieee802154: remove struct padding in context

This change re-orders attributes in struct ieee802154_context to
completely optimize struct padding away. This is done in a way that does
not impact readability.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

fgrandel commented:
@tbursztyka I addressed all your concerns except for creating a separate PR for the IP subsystem - see my comments. It would be great if you could answer my questions in the unresolved comments and/or approve the PR.

fgrandel commented:
@tbursztyka Thanks for your valuable feedback and approval.

@rlubos @jukkar Another approval seems to be required to unblock this PR. You already posted LGTM, seems like its just a missing button press?
